src/libsyntax/ext/derive.rs

```diff
@@ -17,8 +17,11 @@ pub fn collect_derives(cx: &mut ExtCtxt<'_>, attrs: &mut Vec<ast::Attribute>) ->
             return true;
         }
         if !attr.is_meta_item_list() {
-            cx.span_err(attr.span,
-                        "attribute must be of the form `#[derive(Trait1, Trait2, ...)]`");
```

Comment:
The output was supposed to be identical to the generic attribute input validation infra in `feature_gate.rs`.
(Some attributes like `cfg` or `derive` are very special, and have to be checked before that generic infra works.)

Reply:
I removed plenty of attribute-specific messages like "missing traits to be derived" in https://github.com/rust-lang/rust/pull/57321 because the generic messages are good enough and intuitive.

Reply:
The `cfg_attr` check should probably follow the same template.
I still hope that the special attributes can be made less special and go through the generic infra one day.

Reply:
I was considering (but obviously didn't do it in this PR) moving the generic errors to this new format. Is the concern about these being inconsistent or fundamentally against the new output? If it is the former, should I just do that work here for you to see what it'd look like?

Reply:
Fist, consistency, second, I don't like the new wording.
"Bad attribute" doesn't says that the issue is with the attribute input specifically, and not with name or something, the format suggestion is moved too far into a note.
The status quo looks good to me (that's not surprising, since I wrote it) and I don't understand the motivation behind the change.

If I'd change anything, I'd rather move the "must be of the form #[...]" to a suggestion, since it's indeed an approximate suggesion (IIRC, it will be rendered as a label then and will stay close enough).
For the primary message in that case I'd choose something mentioning input, e.g. "ill-formed attribute input".

Reply:
I made these changes and it ballooned the size of the PR, I hope that's ok.

Reply:
FWIW, you can move the EOF staff into a separate PR and consider it r+'d.
I'll be mostly unavailable in the next few days, so this PR may land with a delay.

Reply:
Kept the EOF stuff here and opened a new one with the rewording at https://github.com/rust-lang/rust/pull/61140